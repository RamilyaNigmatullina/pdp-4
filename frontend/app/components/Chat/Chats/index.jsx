import React from 'react';
import PropTypes from 'prop-types';

class Chats extends React.Component {
  renderChat = (chat) => (
    <div key={chat.id}>
      <a href="#" className="chat-link" onClick={(e) => this.props.onClick(e, chat.id)}>
        { chat.interlocutor.full_name }
      </a>
    </div>
  )

  render() {
    return (
      <div className="chats-list">
        { this.props.chats.map((chat) => this.renderChat(chat)) }
      </div>
    );
  }
}

Chats.propTypes = {
  chats: PropTypes.array.isRequired,
  onClick: PropTypes.func.isRequired,
};

export default Chats;
