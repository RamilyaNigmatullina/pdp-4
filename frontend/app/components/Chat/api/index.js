import axios from 'axios';

const MESSAGE_FIELDS = ['id', 'text'];

export const createMessage = (url, formData) => {
  formData.append('fields', MESSAGE_FIELDS);

  axios.post(url, formData);
};
