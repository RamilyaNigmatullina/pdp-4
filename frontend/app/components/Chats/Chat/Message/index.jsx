import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

class Message extends React.Component {
  render() {
    const { avatarUrl, message, isCurrentUserMessage } = this.props;
    const messageClasses = classNames('message', {
      'message--current-user': isCurrentUserMessage,
      'message--interlocutor': !isCurrentUserMessage,
    });

    return (
      <div className={messageClasses} id={`message-${message.id}`}>
        { !isCurrentUserMessage && <img className="rounded-circle" src={avatarUrl} width="32" height="32" /> }
        <span className="message__text">{ message.text }</span>
        { isCurrentUserMessage && <img className="rounded-circle" src={avatarUrl} width="32" height="32" /> }
      </div>
    );
  }
}

Message.propTypes = {
  avatarUrl: PropTypes.string.isRequired,
  isCurrentUserMessage: PropTypes.bool.isRequired,
  message: PropTypes.object.isRequired,
};

export default Message;
