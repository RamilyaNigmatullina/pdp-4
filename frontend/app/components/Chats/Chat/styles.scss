.messages {
  display: flex;
  flex: 1 1 auto;
  flex-direction: column-reverse;
  margin-top: 10px;
  overflow-y: scroll;
  padding: 0 10px;
}

.messages :first-child {
  margin-bottom: 0;
}
