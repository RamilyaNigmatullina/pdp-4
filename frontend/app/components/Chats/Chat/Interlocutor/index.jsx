import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import styles from './styles.module.scss';

class Interlocutor extends React.Component {
  render() {
    const { interlocutor } = this.props;

    return (
      <div className={styles.interlocutor}>
        <div className={styles.interlocutorAvatar}>
          <img
            alt= ""
            className={classNames(styles.image, 'rounded-circle')}
            height="32"
            src={interlocutor.avatar}
            width="32"
          />
        </div>
        <div className={styles.interlocutorFullName}>
          { interlocutor.full_name }
        </div>
      </div>
    );
  }
}

Interlocutor.propTypes = {
  interlocutor: PropTypes.object.isRequired,
};

export default Interlocutor;
