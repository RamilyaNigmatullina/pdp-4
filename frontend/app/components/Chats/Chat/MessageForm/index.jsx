import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import $ from 'jquery';
import 'form-serializer';
import { createMessage } from '../../api/index';
import styles from './styles.module.scss';

class MessageForm extends React.Component {
  constructor(props) {
    super(props);
    this.textInput = React.createRef();

    this.state = { value: this.getValue() };
  }

  componentDidUpdate(prevProps) {
    if (prevProps.chatId !== this.props.chatId) {
      this.setState({ value: this.getValue() });
    }
  }

  renderTextarea() {
    return (
      <div className={classNames(styles.formText, 'form-group', 'text')}>
        <textarea
          className="form-control"
          name="message[text]"
          onChange={(event) => { this.handleChange(event); }}
          onKeyUp={this.handleKeyUp}
          placeholder="Type text..."
          ref={this.textInput}
          rows="1"
          value={this.state.value}
        />
      </div>
    );
  }

  render() {
    return (
      <form
        className={classNames(styles.messageForm, 'simple_form')}
        method="post"
        action={`/api/v1/chats/${this.props.chatId}/messages`}
        onSubmit={this.handleSubmit}
      >

        { this.renderTextarea() }
        <button className={classNames(styles.formButton, 'btn', 'btn-primary')}>Send</button>
      </form>
    );
  }

  handleSubmit = (event) => {
    event.preventDefault();

    if (!this.textInput.current.value) return;

    const formData = $(event.target).serializeObject();
    createMessage(this.props.chatId, formData)
      .then(() => {
        localStorage.setItem(this.chatUid(), '');
        this.setState({ value: '' });
      });
  }

  handleKeyUp = (event) => {
    if (event.keyCode === 13) this.handleSubmit(event);
  }

  handleChange = (event) => {
    const { value } = event.target;

    localStorage.setItem(this.chatUid(), value);
    this.setState({ value });
  }

  chatUid = () => {
    const { chatId, currentUser } = this.props;

    return `chat_${chatId}_${currentUser.id}`;
  }

  getValue = () => localStorage.getItem(this.chatUid()) || ''
}

MessageForm.propTypes = {
  chatId: PropTypes.number.isRequired,
  currentUser: PropTypes.object.isRequired,
};

export default MessageForm;
