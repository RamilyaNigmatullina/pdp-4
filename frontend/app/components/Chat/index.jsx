import React from 'react';
import PropTypes from 'prop-types';
import Message from './Message';
import MessageForm from './MessageForm';

class Chat extends React.Component {
  renderMessages() {
    const { interlocutor, currentUser, messages } = this.props;

    return messages.map((message) => {
      const isCurrentUserMessage = currentUser.id === message.sender_id;
      const avatarUrl = isCurrentUserMessage ? currentUser.avatar : interlocutor.avatar;

      return <Message
        avatarUrl={avatarUrl}
        isCurrentUserMessage={isCurrentUserMessage}
        key={message.id}
        message={message}
      />;
    });
  }

  render() {
    return (
      <div className="chat">
        {this.props.interlocutor.full_name}

        <div className="messages">
          {this.renderMessages()}
        </div>

        { <MessageForm chatId={this.props.chat.id} /> }
      </div>
    );
  }
}

Chat.propTypes = {
  interlocutor: PropTypes.object,
  currentUser: PropTypes.object,
  messages: PropTypes.array,
  chat: PropTypes.object,
};

export default Chat;
