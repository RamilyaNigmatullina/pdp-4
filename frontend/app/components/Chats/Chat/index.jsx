import React from 'react';
import PropTypes from 'prop-types';
import consumer from 'helpers/consumer';
import InfiniteScrolling from 'components/InfiniteScrolling';
import Message from './Message';
import MessageForm from './MessageForm';
import Interlocutor from './Interlocutor';
import { fetchMessages } from '../api/index';

class Chat extends React.Component {
  state = {
    isLastPage: false,
    messages: [],
    page: 1,
  };

  componentDidMount() {
    this.setState({
      subscription: this.subscribeToChannel(),
    });
  }

  componentDidUpdate(prevProps) {
    if (prevProps.chat.id !== this.props.chat.id) {
      this.setState({
        isLastPage: false,
        messages: [],
        page: 1,
        subscription: this.subscribeToChannel(),
      });
    }
  }

  subscribeToChannel = () => {
    const { subscription } = this.state;

    if (subscription) subscription.unsubscribe();

    return consumer.subscriptions.create({
      channel: 'ChatChannel',
      chat_id: this.props.chat.id,
    }, {
      received: this.handleReceived,
    });
  }

  renderMessage = (message) => {
    const { currentUser, chat } = this.props;
    const isCurrentUserMessage = this.isCurrentUserMessage(message);
    const avatarUrl = isCurrentUserMessage ? currentUser.avatar : chat.interlocutor.avatar;

    return (
      <Message
        avatarUrl={avatarUrl}
        isCurrentUserMessage={isCurrentUserMessage}
        key={message.id}
        message={message}
    />);
  }

  render() {
    const { messages, isLastPage } = this.state;
    const { chat } = this.props;

    return (
      <>
        { <Interlocutor interlocutor={chat.interlocutor} /> }

        <div className="messages">
          { <InfiniteScrolling
              isLastPage={isLastPage}
              items={messages}
              loadItems={this.handleLoadMessages}
              renderItem={this.renderMessage}
            /> }
        </div>

        { <MessageForm chatId={chat.id} /> }
      </>
    );
  }

  handleReceived = (message) => {
    if (this.isCurrentUserMessage(message)) this.handleMessageReceived(message);

    this.setState(({ messages }) => ({ messages: [message, ...messages] }), () => {
      const messageElement = document.getElementById(`message-${message.id}`);
      messageElement.scrollIntoView({ behavior: 'smooth' });
    });
  }

  handleMessageReceived = (message) => {
    const chat = { ...this.props.chat, last_message: message, unread_messages_count: 0 };
    const { onMessageReceived } = this.props;

    onMessageReceived(chat);
  }

  handleLoadMessages = () => {
    fetchMessages(this.props.chat.id, this.state.page)
      .then((data) => {
        this.setState(({ messages, page }) => ({
          isLastPage: !data.length,
          messages: [...messages, ...data],
          page: page + 1,
        }));
      });
  }

  isCurrentUserMessage = (message) => this.props.currentUser.id === message.sender_id
}

Chat.propTypes = {
  chat: PropTypes.object.isRequired,
  currentUser: PropTypes.object.isRequired,
  onMessageReceived: PropTypes.func.isRequired,
};

export default Chat;
