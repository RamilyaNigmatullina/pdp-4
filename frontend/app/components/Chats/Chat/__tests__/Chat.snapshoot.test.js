import React from 'react';
import { shallow } from 'enzyme';
import Chat from '../index';

describe('Chat', () => {
  let props;

  const render = () => shallow(<Chat {...props} />);

  beforeEach(() => {
    props = {
      chat: {
        id: 1,
        unread_messages_count: 0,
        interlocutor: {
          id: 2,
          full_name: 'Kevin Doe',
          avatar: '/kevin_doe_avatar.jpg',
        },
        last_message: {
          id: 1,
          text: 'Hello! How are you?',
          sender_id: 1,
          created_at: '2020-03-15T17:39:18.327Z',
        },
      },
      currentUser: {
        id: 1,
        full_name: 'John Smith',
        avatar: '/john_smith_avatar.jpg',
      },
      onMessageReceived: jest.fn(),
    };
  });

  it('renders chat', () => {
    const component = render();

    expect(component).toMatchSnapshot();
  });
});
