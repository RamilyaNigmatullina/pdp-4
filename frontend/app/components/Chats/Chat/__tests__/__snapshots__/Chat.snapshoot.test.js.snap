// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat renders chat 1`] = `
<Fragment>
  <Interlocutor
    interlocutor={
      Object {
        "avatar": "/kevin_doe_avatar.jpg",
        "full_name": "Kevin Doe",
        "id": 2,
      }
    }
  />
  <div
    className="messages"
  >
    <InfiniteScrolling
      isLastPage={false}
      items={Array []}
      loadItems={[Function]}
      renderItem={[Function]}
    />
  </div>
  <MessageForm
    chatId={1}
  />
</Fragment>
`;
